### PR TITLE
Add first external only queue decorator

### PR DESCRIPTION
**Pull request by alexander-schranz**
If you for example crawl https://example.org you want to know the status of a external url e.g https://test.com  but you don't want that test.com is also crawled.

Not sure if thats the correct place for the check another way would be create something like OnlyDescendantOrSelfQueue but checking for referer:

```diff
diff --git a/lib/Model/Queue/OnlyDescendantOrSelfQueue.php b/lib/Model/Queue/OnlyDescendantOrSelfQueue.php
index 75f6aea..a8f2cac 100644
--- a/lib/Model/Queue/OnlyDescendantOrSelfQueue.php
+++ b/lib/Model/Queue/OnlyDescendantOrSelfQueue.php
@@ -33,7 +33,7 @@ class OnlyDescendantOrSelfQueue implements UrlQueue

     public function enqueue(Url $url): void
     {
-        if (false === $url->equalsOrDescendantOf($this->baseUrl)) {
+        if (false === $url->referrer()->equalsOrDescendantOf($this->baseUrl)) {
             return;
         }

```

### CODE REVIEW

**Comment by dantleech** [comment]
Thanks, was thinking about this. Currently this sort of thing is done via. Queue decorators: https://github.com/dantleech/fink/tree/master/lib/Model/Queue which decide.

I guess a new decorator could do something like:

```
public functoin enqueue(Url $url)
{
    if (!$referringUrl = $url->referrer()) {
         // this is the "root" URL... nothing to do here
         $this->innerQueue->enqueue($url);
         return;
    }

    if (!$previousReferringUrl = $url->referrer()->referrer()) {
        // this was a direct child of the main URL, so all good, we can let it into our queue
        $this->innerQueue->enqueue($url);
        return;
    }

    if ($referringUrl->getHost() !== $previousReferringUrl->getHost()) {
        // this is one step too far !?
        return;
    } 

    $this->innerQueue->enqueue($url);
}
```

No idea if that works... if it does then it can be built into the `DispatcherBuilder` with the others and added to the command.

**Comment by dantleech** [comment]
awesome, looks good. could you add a test? there are examples for the other queue decorators :)

File: lib/Model/DispatcherBuilder.php [hunk 1]
```diff
@@ -40,6 +41,11 @@ class DispatcherBuilder
      */
     private $descendantsOnly;
 
+    /**
+     * @var bool
+     */
+    private $firstExternalONly;
```
Comment (dantleech) [inline lib/Model/DispatcherBuilder.php]: typo

File: lib/Command/CrawlCommand.php [hunk 2]
```diff
@@ -54,6 +55,7 @@ protected function configure()
         $this->addOption(self::OPT_OUTPUT, 'o', InputOption::VALUE_REQUIRED, 'Output file');
         $this->addOption(self::OPT_NO_DEDUPE, 'D', InputOption::VALUE_NONE, 'Do not de-duplicate URLs');
         $this->addOption(self::OPT_DESCENDANTS_ONLY, 'l', InputOption::VALUE_NONE, 'Only crawl descendants of the given path');
+        $this->addOption(self::OPT_FIRST_EXTERNAL_ONLY, 'f', InputOption::VALUE_NONE, 'Only crawl the first external not its child');
```
Comment (BackEndTea) [inline lib/Command/CrawlCommand.php]: I'm not sure if `-f` would be the best option here. Generally i'd expect `-f` to be `--force`.

Comment (dantleech) [reply lib/Command/CrawlCommand.php]: maye `x`? but thinking of just surveying existing tools (wget, curl, that python linkchecker) to find a best fit for these options

**Comment by dantleech** [comment]
Thanks @alexander-schranz , will update the changelog. I wonder if this could be generalized a bit more, e.g. `--travel-distance` to describe the maximum number of steps to travel _outside_ of the base URL (as opposed to `--distance` which descibes the number of steps within the URL.

**Pull request merged**